cpp/include/raft/cluster/detail/kmeans.cuh

```diff
@@ -19,967 +19,1056 @@
+  auto n_clusters = params.n_clusters;
+
+  // number of seeding trials for each center (except the first)
+  auto n_trials = 2 + static_cast<int>(std::ceil(log(n_clusters)));
+
+  /*RAFT_LOG_INFO(
+    "Run sequential k-means++ to select %d centroids from %d input samples "
+    "(%d seeding trials per iterations)",
+    n_clusters,
+    n_samples,
+    n_trials);*/
+
+  auto dataBatchSize = getDataBatchSize(params, n_samples);
+
+  // temporary buffers
+  std::vector<DataT> h_wt(n_samples);
+  auto centroidCandidates = raft::make_device_matrix<DataT>(n_trials, n_features, stream);
+  auto costPerCandidate   = raft::make_device_vector<DataT>(n_trials, stream);
+  auto minClusterDistance = raft::make_device_vector<DataT>(n_samples, stream);
+
+  rmm::device_uvector<DataT> distBuffer(n_trials * n_samples, stream);
+  rmm::device_uvector<DataT> L2NormBuf_OR_DistBuf(0, stream);
+  rmm::device_scalar<DataT> clusterCost(stream);
+  rmm::device_scalar<cub::KeyValuePair<int, DataT>> minClusterIndexAndDistance(stream);
+
+  // L2 norm of X: ||c||^2
+  auto L2NormX = raft::make_device_vector<DataT>(n_samples, stream);
+
+  if (metric == raft::distance::DistanceType::L2Expanded ||
+      metric == raft::distance::DistanceType::L2SqrtExpanded) {
+    raft::linalg::rowNorm(
+      L2NormX.data(), X.data(), X.extent(1), X.extent(0), raft::linalg::L2Norm, true, stream);
+  }
 
-    // Transfer result to global memory
-    dists[i] = dist_min;
-    codes[i] = code_min;
+  std::mt19937 gen(params.seed);
+  std::uniform_int_distribution<> dis(0, n_samples - 1);
+
+  // <<< Step-1 >>>: C <-- sample a point uniformly at random from X
+  auto initialCentroid =
+    raft::make_device_matrix_view<const DataT>(X.data() + dis(gen) * n_features, 1, n_features);
+  int n_clusters_picked = 1;
+
+  // reset buffer to store the chosen centroid
+  centroidsRawData.resize(initialCentroid.size(), stream);
+  raft::copy(centroidsRawData.begin(), initialCentroid.data(), initialCentroid.size(), stream);
+
+  //  C = initial set of centroids
+  auto centroids = raft::make_device_matrix_view<DataT>(
+    centroidsRawData.data(), initialCentroid.extent(0), initialCentroid.extent(1));
+  // <<< End of Step-1 >>>
+
+  // Calculate cluster distance, d^2(x, C), for all the points x in X to the nearest centroid
+  minClusterDistanceCompute<DataT, IndexT>(handle,
+                                           params,
+                                           X,
+                                           centroids,
+                                           minClusterDistance.view(),
+                                           L2NormX.view(),
+                                           L2NormBuf_OR_DistBuf,
+                                           workspace,
+                                           metric,
+                                           stream);
+
+  // RAFT_LOG_INFO(" k-means++ - Sampled %d/%d centroids", n_clusters_picked, n_clusters);
```

Comment:
Use `RAFT_LOG_DEBUG` or `RAFT_LOG_TRACE` here if this is something we should keep around for debugging.